packages/debug/src/browser/debug-contribution-manager.ts

```diff
@@ -0,0 +1,247 @@
+        let snippets = await this.debugService.getConfigurationSnippets();
+
+        for (const contributor of this.pluginContributors.values()) {
+            const contribSnippets = await contributor.getConfigurationSnippets();
+            if (contribSnippets) {
+                snippets = snippets.concat(contribSnippets);
+            }
+        }
+
+        return snippets;
+    }
+
+    async create(config: DebugConfiguration): Promise<string> {
+        const contributor = this.pluginContributors.get(config.type);
+        if (contributor) {
+            return contributor.createDebugSession(config);
+        } else {
+            return this.debugService.createDebugSession(config);
+        }
+    }
+
+    async stop(debugType: string, sessionId: string): Promise<void> {
+        const contributor = this.pluginContributors.get(debugType);
+        if (contributor) {
+            return contributor.terminateDebugSession(sessionId);
+        } else {
+            return this.debugService.terminateDebugSession(sessionId);
+        }
+    }
+
+    private async isContributorRegistered(debugType: string): Promise<boolean> {
+        const registeredTypes = await this.debugTypes();
+        return registeredTypes.indexOf(debugType) !== -1;
+    }
+}
+
+/**
+ * Session factory for a client debug session that communicates with debug adapter contributed as plugin.
+ * The main difference is to use a connection factory that creates [IWebSocket](#IWebSocket) over Rpc channel.
+ */
+class PluginDebugSessionFactory implements DebugSessionFactory {
+    constructor(
+        protected readonly terminalService: TerminalService,
+        protected readonly editorManager: EditorManager,
+        protected readonly breakpoints: BreakpointManager,
+        protected readonly labelProvider: LabelProvider,
+        protected readonly messages: MessageClient,
+        protected readonly outputChannelManager: OutputChannelManager,
+        protected readonly debugPreferences: DebugPreferences,
+        protected readonly contributor: DebugPluginContributor
+    ) { }
+
+    get(sessionId: string, options: DebugSessionOptions): DebugSession {
+        let traceOutputChannel: OutputChannel | undefined;
+
+        if (this.debugPreferences['debug.trace']) {
+            traceOutputChannel = this.outputChannelManager.getChannel('Debug adapters');
+        }
+
+        const connection = new DebugSessionConnection(
+            sessionId,
+            this.contributor.getConnectionFactory,
+            traceOutputChannel);
+
+        return new DebugSession(
+            sessionId,
+            options,
+            connection,
+            this.terminalService,
+            this.editorManager,
+            this.breakpoints,
+            this.labelProvider,
+            this.messages,
+            traceOutputChannel,
+        );
+    }
+}
+
+/**
+ * Describes what plugin contribution has to provide for a client.
+ */
+export interface DebugPluginContributor {
```

Comment:
debug extension should not have notion of plugins

plugins should be encapsulated into `@theia/plugin-ext`